src/Hypsography/Hypsography.jl

```diff
@@ -15,42 +24,88 @@ using StaticArrays, LinearAlgebra
 
 
 """
-    ref_z_to_physical_z(adaption::HypsographyAdaption, z_ref, z_surface, z_top)
+    ref_z_to_physical_z(adaption::HypsographyAdaption, z_ref::ZPoint, z_surface::ZPoint, z_top::ZPoint) :: ZPoint
 
 Convert reference `z`s to physical `z`s as prescribed by the given adaption.
+
+This function has to be the inverse of `physical_z_to_ref_z`.
 """
 function ref_z_to_physical_z(
     adaption::HypsographyAdaption,
-    z_ref,
-    z_surface,
-    z_top,
+    z_ref::Geometry.ZPoint,
+    z_surface::Geometry.ZPoint,
+    z_top::Geometry.ZPoint,
 ) end
 
-# For no adaption, z_physical is z_ref.
-function ref_z_to_physical_z(adaption::Nothing, z_ref, z_surface, z_top)
-    return z_ref
-end
-
 """
-    _check_adaption_constraints(adaption, z_ref, z_surface, z_top)
+    physical_z_to_ref_z(adaption::HypsographyAdaption, z_ref::ZPoint, z_surface::ZPoint, z_top::ZPoint) :: ZPoint
 
-Ensures that the `adaption` is well defined for the given parameters.
+Convert physical `z`s to reference `z`s as prescribed by the given adaption.
+
+This function has to be the inverse of `ref_z_to_physical_z`.
 """
-function _check_adaption_constraints(adaption, z_ref, z_surface, z_top) end
+function physical_z_to_ref_z(
+    adaption::HypsographyAdaption,
+    z_phys::Geometry.ZPoint,
+    z_surface::Geometry.ZPoint,
+    z_top::Geometry.ZPoint,
+) end
+
+# Flat, z_ref = z_physical
+
+function ref_z_to_physical_z(
+    ::Flat,
+    z_ref::Geometry.ZPoint,
+    z_surface::Geometry.ZPoint,
+    z_top::Geometry.ZPoint,
+)
+    return z_ref
+end
+
+function physical_z_to_ref_z(
+    ::Flat,
+    z_physical::Geometry.ZPoint,
+    z_surface::Geometry.ZPoint,
+    z_top::Geometry.ZPoint,
+)
+    return z_physical
+end
 
 """
     LinearAdaption(surface::Field)
 
 Locate the levels by linear interpolation between the surface field and the top
 of the domain, using the method of [GalChen1975](@cite).
 """
-struct LinearAdaption{F <: Union{Fields.Field, Nothing}} <: HypsographyAdaption
-    # Union can be removed once deprecation removed.
+struct LinearAdaption{F <: Fields.Field} <: HypsographyAdaption
     surface::F
+    function LinearAdaption(surface::Fields.Field)
+        if eltype(surface) <: Real
+            @warn "`LinearAdaptation`: `surface` argument scalar field has been deprecated. Use a field `ZPoint`s."
+            surface = Geometry.ZPoint.(surface)
+        end
+        new{typeof(surface)}(surface)
+    end
 end
 
-function ref_z_to_physical_z(adaption::LinearAdaption, z_ref, z_surface, z_top)
-    return z_ref + (1 - z_ref / z_top) * z_surface
+# This method is invoked by the ExtrudedFiniteDifferenceGrid constructor
+function ref_z_to_physical_z(
+    ::LinearAdaption,
+    z_ref::Geometry.ZPoint,
+    z_surface::Geometry.ZPoint,
+    z_top::Geometry.ZPoint,
+)
+    Geometry.ZPoint(z_ref.z + (1 - z_ref.z / z_top.z) * z_surface.z)
```

Comment:
I was able to reproduce the type issue : 
```
julia> grid = Grids.ExtrudedFiniteDifferenceGrid(horz_grid,
                                                     vert_grid,
                                                     warp;
                                                     )
ERROR: MethodError: no method matching ref_z_to_physical_z(::ClimaCore.Hypsography.LinearAdaption{…}, ::ClimaCore.DataLayouts.VF{…}, ::ClimaCore.DataLayouts.IJFH{…}, ::ClimaCore.Geometry.ZPoint{…})

Closest candidates are:
  ref_z_to_physical_z(::ClimaCore.Hypsography.LinearAdaption, ::ClimaCore.Geometry.ZPoint, ::ClimaCore.Geometry.ZPoint, ::ClimaCore.Geometry.ZPoint)
   @ ClimaCore ~/Codes/ClimaCore.jl/src/Hypsography/Hypsography.jl:92
  ref_z_to_physical_z(::ClimaCore.Grids.HypsographyAdaption, ::ClimaCore.Geometry.ZPoint, ::ClimaCore.Geometry.ZPoint, ::ClimaCore.Geometry.ZPoint)
``` 
since the inputs in lines `180-184` result in DataLayouts `VF` and `IJFH`, rather than `ZPoints`.  The current tests, as long as they assume `CPUSingleThreaded` for the `terrain_warp` component will not catch this. 

```
function ref_z_to_physical_z(
    ::LinearAdaption,
    z_ref,
    z_surface,
    z_top,
)
    Geometry.ZPoint.(z_ref.z .+ (1 .- z_ref.z ./ z_top.z) .* z_surface.z)
end
``` fixes this by removing the type constraint (but we should prescribe the appropriate ZPoint DataLayout types here.)

Reply:
```
2D Extruded Terrain Warped Space: Error During Test at /home/asridhar/Codes/ClimaCore.jl/test/Spaces/terrain_warp.jl:207
  Got exception outside of a @test
  MethodError: no method matching ref_z_to_physical_z(::ClimaCore.Hypsography.LinearAdaption{ClimaCore.Fields.Field{ClimaCore.DataLayouts.IFH{ClimaCore.Geometry.ZPoint{Float32}, 3, Array{Float32, 3}}, ClimaCore.Spaces.SpectralElementSpace1D{ClimaCore.Grids.SpectralElementGrid1D{ClimaCore.Topologies.IntervalTopology{ClimaComms.SingletonCommsContext{ClimaComms.CPUSingleThreaded}, ClimaCore.Meshes.IntervalMesh{ClimaCore.Domains.IntervalDomain{ClimaCore.Geometry.XPoint{Float32}, Nothing}, LinRange{ClimaCore.Geometry.XPoint{Float32}, Int64}}, @NamedTuple{}}, ClimaCore.Quadratures.GLL{3}, ClimaCore.Geometry.CartesianGlobalGeometry, ClimaCore.DataLayouts.IFH{ClimaCore.Geometry.LocalGeometry{(1,), ClimaCore.Geometry.XPoint{Float32}, Float32, StaticArraysCore.SMatrix{1, 1, Float32, 1}}, 3, Array{Float32, 3}}, ClimaCore.DataLayouts.IFH{Float32, 3, Array{Float32, 3}}}}}}, ::ClimaCore.DataLayouts.VF{ClimaCore.Geometry.ZPoint{Float32}, SubArray{Float32, 2, Matrix{Float32}, Tuple{Base.Slice{Base.OneTo{Int64}}, UnitRange{Int64}}, true}}, ::ClimaCore.DataLayouts.IFH{ClimaCore.Geometry.ZPoint{Float32}, 3, Array{Float32, 3}}, ::ClimaCore.Geometry.ZPoint{Float32})
```